Add EIP-5075: RateLimit, an outflow limiter for assets (#5075)
ethereum/EIPs: EIPS/EIPs/eip-5075.md

## Backwards Compatibility
This will be non backwards compatible with contracts that are NOT whitelisted which assume they can just withdraw large positions, if those positions are large relative to the target contracts net position by all users. Should this be the case the contract can be fully compatible by simply being whitelisted, skipping all rate limit checks.

## Reference Implementation

```
  // SPDX-License-Identifier: CC0
pragma solidity ^0.8;

import"./IERC20.sol";

contract rateLimit {
    mapping(address => TokenInfo) public infoL;
    struct TokenInfo {
        uint32 timestamp; //tracks most recent send
        uint224 rate; //tracks amount sent for the last timeLimit time
    }

>     uint256 public timeLimit = 3600; //how long in seconds to limit within, recommend 1h = 3600

Reviewer: nit: maybe `timeWindow`? it's referred to this way in most explaining comments

Author: Will do

>     uint16 public rateLimit = 1000; //the basis points (00.0x%) to allow as the max sent within the last timeLimit time

Reviewer: nit: `rate` is in amount unit, but `rateLimit` is in bps which is a bit confusing to read. Maybe rename `rate` to `outflow` (because an amount is not a rate?)

Author: Noted, will do

        uint256 amount,
        address token
    ) public returns (bool success) {
        TokenInfo storage info = infoL[token];

        if (whitelisted[msg.sender] == 1) {
            if (address(token) == address(0)) {
                payable(to).transfer(amount);
                return(true);
            } else {
                return(IERC20(token).transfer(to, amount));
            }
        }
        //used if the asset is ETH and not an ERC20 token
        else {
            if (address(token) == address(0)) {
                //used to get around solidity 0.8 reverts
                unchecked {
                    //checks to see if the last transaction was outside the time window to track outflow for limiting
                    if (timeLimit <= block.timestamp - info.timestamp) {
                        info.rate = 0;
                    }
                    //if the last transaction was within the time window, decreases the tracked outflow rate relative to the time elapsed, so that the limit is able to update in realtime rather than in blocks, making flows smooth, and increasing the rate available as time increases without a transaction
                    else {

>                         info.rate -= uint224(

Reviewer: I think this may revert due to underflow if balance increased by a lot since the last rate was recorded, or if rateLimit or timeLimit paremeters changed in a way to case underflow. Should probably check for this before trying to subtract.

Author: Will do

                                (timeLimit /
                                    (block.timestamp - info.timestamp)) /
                                10000
                        );
                    }
                }
                //increases the tracked rate for the current time window by the amount sent out
                info.rate += uint224(amount);
                //revert if the outflow exceeds rate limit
                require(
                    info.rate <= (rateLimit * address(this).balance) / 10000
                );
                //sets the current time as the last transfer for the token
                info.timestamp = uint32(block.timestamp);
                //transfers out
                payable(to).transfer(amount);
                return(true);
                //if the token is a ERC20 token
            } else {
                //used to get around solidity 0.8 reverts
                unchecked {
                    //checks to see if the last transaction was outside the time window to track outflow for limiting
                    if (timeLimit <= block.timestamp - info.timestamp) {
                        info.rate = 0;
                    }

>                     //if the last transaction was within the time window, decreases the tracked outflow rate relative to the time elapsed, so that the limit is able to update in realtime rather than in blocks, making flows smooth, and increasing the rate available as time increases without a transaction

Reviewer: this is a nice trick, but I think it's equivalent to spreading the recent outflow over the whole previous time window. Which in effect means that rateLimit as it's phrased can always be exceeded. E.g. if previous outflow was near the end of the window (e.g 59 minutes ago), the new outflow can almost be the size of `rateLimit` (because 59/60 of capacity are added) which means that the `rateLimit` strictly speaking is exceeded by a factor of almost two for the last hour.

I'm not sure if this means that actual rate limit is at most 2X the parameter for it (which is still limited), but probably needs to be mentioned in docs and description of the functionality.

Author: Have updated accordingly, will attach notice about implementation for how to set the and handle the limit being an advised plan not a requirement.

                    else {

>                         info.rate -= uint224(

Reviewer: similar concern about revert due to undeflow due to balance or settings changes

Author: Will do

---
eip: 5075
title: RateLimit: an outflow limiter for assets
description: Limits outflows for all contract assets to a given rate in a given timeframe
author: PR0 (@pr0toshi)
discussions-to: https://ethereum-magicians.org/t/eip-x-limit-token-outflows-within-timeframe-to-limit-hack-losses/9172
status: Draft
type: Standards Track

> category ERC

Reviewer 2: ```suggestion
category: ERC
```

---
eip: 5075
title: rateLimit - An outflow limiter for assets
description: Limits outflows for all contract assets to a given rate in a given timeframe
author: PR0 (@pr0toshi)
discussions-to: https://ethereum-magicians.org/t/eip-x-limit-token-outflows-within-timeframe-to-limit-hack-losses/9172
status: Draft
type: Standards Track
category ERC
created: 2022--05-05
---

## Abstract

> An inheritable implementation for a rate limiter transferL fn which replaces all asset outflow transfer fns to unknown external addresses.  By limiting asset outflows for all assets in a contract within a customisable timeframe, contracts which suffer from hacks will have losses limited to the given rateLimit param, which is dynamic and relative to a contracts net total assets, being tracked at the per asset level.

Author: updated. specified transferL as a new fn to replace standard transfer fns for asset outflows,

## Motivation
The Crypto ecosystem continues to suffer from record setting losses from hacks, which are usually highly concentrated single-transaction executed or flash loan based, or rapidly executed attacks using vulnerabilities in contracts to withdraw  more funds than would be intended. While there are usually mechanisms should a hack be undergoing, hacks cannot be limited or stopped in real time or mid transaction.

This ERC creates a simple to use and plug and play style improvement which limits outflows to the intended expected limits, and reverts should these limits be exceeded. This will ultimately limit the impoact for almost all vulnerabilities in contracts which find vulnerabilities in contracts and then use the contracts native withdrawal mechanism to withdraw and drain funds. The fn has been optimized to have a low overhead on transfers and is intended to replace transfer fns throughout contracts for any publicly accressible transfer to external addresses, which are the most common route for asset withdrawal post hack.

## Specification

> # Methods

Reviewer 2: ```suggestion
### Methods
```

wrong heading level here

## Security Considerations
The above discussion shows why we decided to implement the per token tracking and why and how this should be used to help limit losses from hacks. Contracts must consider any whitelists very carefully as these may be able to be used to drain assets by themselves having been hacked or through non intended use. Teams and projects must also consider their usual outflows based on analytics to try and not impact user experience, balancing both the limit in potential damage from the hack as well as the user experience from transaction reverts. The getLimitLeft can be used in the front end to let users know that there is high demand and to wait for outflows from the contracts. This also does not affect approvals and transferFrom to be used on the contracts balances, so contracts must be aware and manage approvals carefully as these are not checked, using transferL where possible.
  
Note rateLimit can be set to 1000 (100% in basis points) to allow for the feature to be effectively disabled. Had considered doing a flag and a check but this would create a slight gas inefficiency on all transactions during usual operation.

> 

Reviewer 2: Missing: rationale, security considerations, backwards compatibility sections

Author: Uh all 3 are there at l64.

Author: unless they are preferable to be under the reference impl

Author: the EIP1 template states that its right as shown.

Reviewer 2: @pr0toshi sorry my mistake, the spacing made it difficult to see the `## Rationale`. Can you please add some spacing between sections? You are also still missing security considerations.

---
eip: 5075

> title: RateLimit: an outflow limiter for assets

Reviewer 2: ```suggestion
title: "RateLimit: an outflow limiter for assets"
```

> title: "RateLimit: an outflow limiter for assets"

Reviewer 3: ```suggestion
title: RateLimit: an outflow limiter for assets
```

description: Limits outflows for all contract assets to a given rate in a given timeframe
author: PR0 (@pr0toshi)
discussions-to: https://ethereum-magicians.org/t/eip-x-limit-token-outflows-within-timeframe-to-limit-hack-losses/9172
status: Draft
type: Standards Track
category: ERC

> created: 2022--05-05

Reviewer 3: ```suggestion
created: 2022-05-05
```

---

## Abstract

> An inheritable implementation for a rate limiter using a new transferL fn which replaces all asset outflow transfer fns to unknown external addresses. By limiting asset outflows for all assets in a contract within a customisable timeframe, contracts which suffer from hacks will have losses limited to the given rateLimit param, which is dynamic and relative to a contracts net total assets, being tracked at the per asset level.

Reviewer 3: ```suggestion
An inheritable implementation for a rate limiter using a new `transferL` function which replaces all asset outflow transfer functions to unknown external addresses. By limiting asset outflows for all assets in a contract within a customizable timeframe, contracts that suffer from hacks will have losses limited to the given `rateLimit` parameter, which is dynamic and relative to a contract's net total assets, being tracked at the per asset level.
```

I personally would also prefer if the function name was more descriptive, say `transferWithLimit` or `transferLimited`.

> Limiting outflows within given timeframes to external addresses stops both single transaction vulnerabilities which drain all funds, aswell as those which are not single transaction/flash loan based and may be performed throughout a given timeframe. By delaying and limiting outflows, teams will have time to respond with the security mechanisms in contracts determined by the teams (upgrades, freezes, not rate limited onlyOwner multisig based withdrawals), limiting losses by users. Includes optional whitelist functionality.

Reviewer 3: ```suggestion
Limiting outflows within given timeframes to external addresses stops both single transaction vulnerabilities which drain all funds, as well as those which are not single transaction/flash loan based and may be performed throughout a given timeframe. By delaying and limiting outflows, teams will have time to respond with the security mechanisms in contracts determined by the teams (upgrades, freezes, not rate limited `onlyOwner` multisig based withdrawals), limiting losses by users.

This EIP includes an optional whitelist functionality.
```

## Motivation

> The Crypto ecosystem continues to suffer from record setting losses from hacks, which are usually highly concentrated single-transaction executed or flash loan based, or rapidly executed attacks using vulnerabilities in contracts to withdraw  more funds than would be intended. While there are usually mechanisms should a hack be undergoing, hacks cannot be limited or stopped in real time or mid transaction.

Reviewer 3: ```suggestion
The cryptocurrency ecosystem continues to suffer from record-setting losses due to attacks, often using single or multi-transaction exploits (with flash loans, for example), to rapidly withdraw more funds than intended. While there are often mechanisms implemented to freeze a contract in the aftermath of an attack, these types of attacks cannot be limited or stopped in real-time or mid-transaction.
```

> 
> This ERC creates a simple to use and plug and play style improvement which limits outflows to the intended expected limits, and reverts should these limits be exceeded. This will ultimately limit the impoact for almost all vulnerabilities in contracts which find vulnerabilities in contracts and then use the contracts native withdrawal mechanism to withdraw and drain funds. The fn has been optimized to have a low overhead on transfers and is intended to replace transfer fns throughout contracts for any publicly accressible transfer to external addresses, which are the most common route for asset withdrawal post hack.

Reviewer 3: ```suggestion
This EIP creates a simple, plug-and-play layer of protection that limits outflows to the intended expected limits, and reverts should these limits be exceeded. This will limit the extent of the damage in the event of a fund-draining vulnerability. The function has been optimized to have minimal overhead on transfers and is intended to replace transfer functions throughout contracts. 
```

> 

Reviewer 3: ```suggestion
```

Remove extra whitespace

## Security Considerations
The above discussion shows why we decided to implement the per token tracking and why and how this should be used to help limit losses from hacks. Contracts must consider any whitelists very carefully as these may be able to be used to drain assets by themselves having been hacked or through non intended use. Teams and projects must also consider their usual outflows based on analytics to try and not impact user experience, balancing both the limit in potential damage from the hack as well as the user experience from transaction reverts. The getLimitLeft can be used in the front end to let users know that there is high demand and to wait for outflows from the contracts. This also does not affect approvals and transferFrom to be used on the contracts balances, so contracts must be aware and manage approvals carefully as these are not checked, using transferL where possible.
  

> Note rateLimit can be set to 1000 (100% in basis points) to allow for the feature to be effectively disabled. Had considered doing a flag and a check but this would create a slight gas inefficiency on all transactions during usual operation.

Reviewer 3: ```suggestion
Note that `rateLimit` may be set to 1000 (100% in basis points), which effectively disables the rate limit. A flag had been considered, but it would result in higher gas fees and therefore was deemed unneccesary.
```

> 

Reviewer 3: ```suggestion
```

Extraneous whitespace

## Copyright

> Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

Reviewer 3: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```
